source/agora/utils/Test.d

```diff
@@ -627,60 +632,360 @@ unittest
+        this.leftover.value.mustAdd(tx.outputs[index].value);
+        return this;
+    }
+
+    /***************************************************************************
+
+        Finalize the transaction, signing the input, and reset the builder
+
+        Params:
+            type = type of `Transaction`
+
+        Returns:
+            The finalized & signed `Transaction`.
+
+    ***************************************************************************/
+
+    public Transaction sign (TxType type = TxType.Payment) @safe
+    {
+        assert(this.inputs.length, "Cannot sign input-less transaction");
+        assert(this.data.outputs.length || this.leftover.value > Amount(0),
+               "Onput-less transactions are not valid");
+
+        this.data.type = type;
+
+        // Finalize the transaction by adding inputs
+        foreach (ref in_; this.inputs)
+            this.data.inputs ~= Input(in_.hash(), in_.index);
+
+        // Add the refund tx, if needed
+        if (this.leftover.value > Amount(0))
+            this.data.outputs ~= this.leftover;
+
+        // Get the hash to sign
+        const txHash = this.data.hashFull();
+        // Sign all inputs using WK keys
+        foreach (idx, ref in_; this.inputs)
+        {
+            auto ownerKP = this.getOwnerKeyPair(in_.output.address);
+            assert(ownerKP !is KeyPair.init,
+                    "Address not found in Well-Known keypairs: "
+                    ~ in_.output.address.toString());
+            this.data.inputs[idx].signature = () @trusted
+                { return ownerKP.secret.sign(txHash[]); }();
+        }
+
+        // Return the result and reset this
+        this.inputs = null;
+        this.leftover = Output.init;
+        scope (success) this.data = Transaction.init;
+        return this.data;
+    }
+
+    /***************************************************************************
+
+        Finds a well-known `KeyPair` corresponding to the `PublicKey`.
+
+        Params:
+            address = finding `PublicKey`
+
+        Returns:
+            returns KeyPair.init if `address` is not found.
 
-    auto remainder = amount.div(toward.length);
-    Transaction result = Transaction(type);
-    foreach (addr; toward)
-        result.outputs ~= Output(amount, addr);
+    ***************************************************************************/
 
-    // Add the remainder to the first output.
-    result.outputs[0].value.mustAdd(remainder);
-    const inputHash = input.hashFull();
+    private KeyPair getOwnerKeyPair (scope const PublicKey address) @safe
+    {
+        // FIXME: Using the binary version didn't work (?)
+        if (address.toString() == "GCOQEOHAUFYUAC6G22FJ3GZRNLGVCCLESEJ2AXBIJ5BJNUVTAERPLRIJ")
+            return  () @trusted {
+                return KeyPair.fromSeed(Seed.fromString(
+                    "SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4"));
+            }();
```

Comment:
Perhaps we can just have the genesis key be in the well-known key (not as a `A`, `B`, `C`, etc... but as `Genesis`).

Reply:
Aha so it's not in the well known key list. Could you add a comment for this in the fixme and replace the `'?'` so it's more obvious?

Reply:
Fixed